kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ingress/ClusterIPIngressDefinition.java

```diff
@@ -0,0 +1,122 @@
+        ClusterIPIngressInstance {
+            Objects.requireNonNull(definition);
+            sanityCheckPortRange(definition, firstIdentityfulPort, lastIdentityfulPort);
+        }
+
+        @Override
+        public VirtualClusterGateway gatewayConfig() {
+            return new VirtualClusterGateway("default",
+                    new PortIdentifiesNodeIdentificationStrategy(new HostPort("localhost", firstIdentityfulPort),
+                            absoluteServiceHost(), null, List.of(new NamedRange("default", 0, NUM_BROKERS - 1))),
+                    null,
+                    Optional.empty());
+        }
+
+        @Override
+        public Stream<ServiceBuilder> services() {
+            var serviceSpecBuilder = new ServiceBuilder()
+                    .withNewMetadata()
+                    .withName(serviceName(definition.cluster, definition.resource))
+                    .withNamespace(namespace(definition.cluster))
+                    .addToLabels(standardLabels(definition.primary))
+                    .addNewOwnerReferenceLike(ResourcesUtil.ownerReferenceTo(definition.primary)).endOwnerReference()
+                    .endMetadata()
+                    .withNewSpec()
+                    .withSelector(ProxyDeployment.podLabels(definition.primary));
+            for (int i = firstIdentityfulPort; i <= lastIdentityfulPort; i++) {
+                serviceSpecBuilder = serviceSpecBuilder
+                        .addNewPort()
+                        .withName(name(definition.cluster) + "-" + i)
+                        .withPort(i)
+                        .withTargetPort(new IntOrString(i))
+                        .withProtocol("TCP")
+                        .endPort();
+            }
+            return Stream.of(serviceSpecBuilder.endSpec());
+        }
+
+        private static void sanityCheckPortRange(ClusterIPIngressDefinition definition, int startPortInc, int endPortInc) {
+            int requiredPorts = definition.identityfulPortsRequired();
+            if ((endPortInc - startPortInc + 1) != requiredPorts) {
+                throw new IllegalArgumentException("require " + requiredPorts + " ports");
+            }
+        }
+
+        @Override
+        public Stream<ContainerPort> proxyContainerPorts() {
+            Stream<ContainerPort> bootstrapPort = Stream.of(new ContainerPortBuilder().withContainerPort(firstIdentityfulPort)
+                    .withName(firstIdentityfulPort + "-bootstrap").build());
+            Stream<ContainerPort> ingressNodePorts = IntStream.range(0, NUM_BROKERS).mapToObj(
+                    nodeId -> {
+                        int port = firstIdentityfulPort + nodeId + 1;
+                        return new ContainerPortBuilder().withContainerPort(port)
+                                .withName(port + "-node").build();
+                    });
+            return Stream.concat(bootstrapPort, ingressNodePorts);
+        }
+
+        String absoluteServiceHost() {
```

Comment:
I've never heard anyone use the term "absolute host name". I have heard the term "(fully) qualified host name".